File: core/src/main/java/fr/skyost/serialkey/core/listener/BunchOfKeysListener.java
package fr.skyost.serialkey.core.listener;

import fr.skyost.serialkey.core.SerialKeyPlugin;

/**
 * A listener that allows to listen bunch of keys related events.
 */

public class BunchOfKeysListener<I, L> extends SerialKeyListener<I, L> {

	/**
	 * Creates a new bunch of keys listener instance.
	 *
	 * @param plugin The plugin.
	 */

	public BunchOfKeysListener(final SerialKeyPlugin<I, L> plugin) {
		super(plugin);
	}

	/**
	 * Triggered when a player right clicks on air (no block).
	 *
	 * @param item The item that was used.
	 * @param cancelIfCreateInventory Cancels the event if the bunch of keys inventory has been successfully created.
	 */

	protected void onPlayerRightClickOnAir(final I item, final Runnable cancelIfCreateInventory) {
		if(!itemManager.isBunchOfKeys(item)) {
			return;
		}

		cancelIfCreateInventory.run();
	}

	/**
	 * Triggered when a player right clicks on a block.
	 *
	 * @param location The block location.
	 * @param item The item that was used.
	 * @param cancelIfCreateInventory Cancels the event if the bunch of keys inventory has been successfully created.
	 */

	protected void onPlayerRightClickOnBlock(final L location, final I item, final Runnable cancelIfCreateInventory) {
		if(padlockManager.hasPadlock(location)) {
			return;
		}

		onPlayerRightClickOnAir(item, cancelIfCreateInventory);
	}

}

File: core/src/main/java/fr/skyost/serialkey/core/listener/BunchOfKeysListener.kt
package fr.skyost.serialkey.core.listener

import fr.skyost.serialkey.core.SerialKeyPlugin

/**
 * A listener that allows to listen bunch of keys related events.
 */
open class BunchOfKeysListener<I, L>(plugin: SerialKeyPlugin<I, L>) : SerialKeyListener<I, L>(plugin) {
    /**
     * Triggered when a player right clicks on air (no block).
     *
     * @param item The item that was used.
     * @param cancelIfCreateInventory Cancels the event if the bunch of keys inventory has been successfully created.
     */
    protected fun onPlayerRightClickOnAir(item: I?, cancelIfCreateInventory: Runnable) {
        if (!itemManager.isBunchOfKeys(item)) {
            return
        }
        cancelIfCreateInventory.run()
    }

    /**
     * Triggered when a player right clicks on a block.
     *
     * @param location The block location.
     * @param item The item that was used.
     * @param cancelIfCreateInventory Cancels the event if the bunch of keys inventory has been successfully created.
     */
    protected fun onPlayerRightClickOnBlock(location: L, item: I?, cancelIfCreateInventory: Runnable) {
        if (padlockManager.hasPadlock(location)) {
            return
        }
        onPlayerRightClickOnAir(item, cancelIfCreateInventory)
    }
}